### ISSUE TITLE
Guacamole and XRDP - Fastpath and other errors.

### ISSUE THREAD

**Issue by MattzFiber** [neutrinolabs/xrdp#1785]
Greetings! 

Thanks for the awesome code, and great software. I've spent many hours digging here and in the xrdp community, and its amazing to see the support for this project. 

I seem to reliably face an issue with a deployment of XRDP and Guacamole, regardless if I use the latest of both, or older versions of both software. 

When connecting to a session, everything works for a few moments, including sound (compiled with guides here), video, and seeminly clipboard. 
If I log in, play a sound (eg a Youtube video or something) and remove the mouse from the RDP window, it will work reliably for a long time. 

When I move the mouse curser into the session, do things, "wiggle the mouse around", I will always have some kind of a crash in the logs, and the sound will no longer work. The RDP session stays connected, and the video continues, but the sound drops and I note error logs each time. 
I have reproduced this on both Ubuntu and CentOS on the xrdp side, with multiple Guacamole versions. I use a minimal X setup, just to allow access to a web-browser that auto starts. 

I am not able to produce this bug when using the built in Windows RDP client, where XRDP works flawlessly for as long as I've tried to test it. 

Current version I have built from this git: 
```
$ xrdp -v
xrdp 0.9.15
  A Remote Desktop Protocol Server.
  Copyright (C) 2004-2020 Jay Sorg, Neutrino Labs, and all contributors.
  See https://github.com/neutrinolabs/xrdp for more information.

  Configure options:
      --enable-jpeg

  Compiled with OpenSSL 1.0.2k-fips  26 Jan 2017
```

Logs from syslog: 
```
Jan 18 23:57:03 gagarin xrdp-sesman[1510]: [INFO ] Xorg :10 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log
Jan 18 23:57:03 gagarin xrdp-sesman[1508]: [INFO ] waiting for window manager (pid 1509) to exit
Jan 18 23:57:03 gagarin xrdp[1505]: [INFO ] lib_mod_log_peer: xrdp_pid=1505 connected to X11rdp_pid=1510 X11rdp_uid=1000 X11rdp_gid=1000 client_ip=10.66.1.6 client_port=43804
Jan 18 23:57:03 gagarin xrdp[1505]: [ERROR] libxrdp_query_channel - Channel 0 name cliprdr
Jan 18 23:57:03 gagarin xrdp-chansrv[1513]: [INFO ] Socket 12: AF_UNIX connection received
Jan 18 23:57:04 gagarin xrdp[1505]: [ERROR] libxrdp_query_channel - Channel 1 name rdpdr
Jan 18 23:57:04 gagarin xrdp[1505]: [ERROR] libxrdp_query_channel - Channel 2 name rdpsnd
Jan 18 23:57:04 gagarin xrdp[1505]: [ERROR] libxrdp_query_channel - Channel out of range 3
Jan 18 23:57:04 gagarin dbus[662]: [system] Activating via systemd: service name='org.freedesktop.RealtimeKit1' unit='rtkit-daemon.service'
Jan 18 23:57:04 gagarin systemd: Starting RealtimeKit Scheduling Policy Service...
Jan 18 23:57:04 gagarin dbus[662]: [system] Successfully activated service 'org.freedesktop.RealtimeKit1'
Jan 18 23:57:04 gagarin systemd: Started RealtimeKit Scheduling Policy Service.
Jan 18 23:57:04 gagarin pulseaudio: sink_process_msg: not running
Jan 18 23:57:04 gagarin pulseaudio: close_send:
Jan 18 23:57:04 gagarin xrdp-chansrv[1513]: [INFO ] sound_process_training: round trip time 101
Jan 18 23:57:30 gagarin pulseaudio: sink_process_msg: running
Jan 18 23:57:30 gagarin xrdp-chansrv[1513]: [INFO ] Socket 16: AF_UNIX connection received
Jan 18 23:57:30 gagarin pulseaudio: Connected ok fd 13
Jan 18 23:57:40 gagarin xrdp[1505]: [WARN ] xrdp_fastpath_process_input_event: unknown eventCode 6
Jan 18 23:57:40 gagarin xrdp[1505]: [WARN ] xrdp_fastpath_process_input_event: unknown eventCode 5
Jan 18 23:57:54 gagarin xrdp[1505]: [ERROR] libxrdp_process_data returned non zero
Jan 18 23:57:54 gagarin xrdp[1505]: [WARN ] xrdp_fastpath_process_input_event: unknown eventCode 5
Jan 18 23:57:54 gagarin xrdp[1505]: [WARN ] xrdp_fastpath_process_input_event: unknown eventCode 5
Jan 18 23:57:54 gagarin xrdp[1505]: [ERROR] libxrdp_process_data returned non zero
Jan 18 23:57:55 gagarin xrdp[1505]: [WARN ] xrdp_fastpath_process_input_event: unknown eventCode 6
Jan 18 23:57:55 gagarin pulseaudio: data_send: send failed
```

If I disable FastPath and set it to "none" from "both", the fastpath mention in the logs is removed, but the symtoms are the same: 

```
Jan 19 00:10:58 gagarin pulseaudio: sink_process_msg: running
Jan 19 00:10:58 gagarin xrdp-chansrv[2153]: [INFO ] Socket 16: AF_UNIX connection received
Jan 19 00:10:58 gagarin pulseaudio: Connected ok fd 13
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 8
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 6
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 6
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 11
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 5
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 6
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 10
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 9
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 11
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 8
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 5
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 13
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 11
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 14
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 1
Jan 19 00:11:00 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 9
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 14
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 14
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:01 gagarin pulseaudio: data_send: send failed
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] xrdp_channel_process returned unhandled error
Jan 19 00:11:01 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 5
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 13
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 10
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 4
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 14
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 10
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 11
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] libxrdp_process_data returned non zero
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 8
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 14
Jan 19 00:11:02 gagarin xrdp[2141]: [ERROR] unknown in libxrdp_process_data: code= 9
```

Again, it doesnt seem to happen using another RDP client, so its absolutely within the realm of possibility this issue is unrelated to XRDP, but perhaps someone might know whats going on. 

I cannot see any other issue noting these kinds of errors.
If there is any other logs or steps to reproduce an error that would be helpful, please let me know. 

Many thanks!

**Comment by MattzFiber** [comment]
FWIW, I put some more time into the Guac side of things and went all the way back to guac:1.0.0, from Jan of 2019. 
This version seems to work, and I cannot create the failure condition above, while 1.1.0 contains the failure. 

Notably, the 1.1.0 release notes : "Migration to FreeRDP 2.0.0"
https://guacamole.apache.org/releases/1.1.0/

**Comment by aquesnel** [comment]
Hi @MattzFiber 

I've recently added a lot more detailed error logging in the still unmerged pull request #1742 that I think might help diagnose this issue.

Could you please try reproducing the error that you are getting and share the logs with xrdp recompiled from the tip of the `logging_add_libxrdp` branch and using the `--enable-xrdpdebug` configure flag? 

I'm interested in both the logs at the error/warn level as well as at the trace level. **WARNING**:  the trace logs with the xrdpdebug configure flag does log sensitive information like **certificate private keys** and **passwords**.

**Comment by purplebadger** [comment]
@MattzFiber did you ever resolve this issue with Guac version 1.0.0+?

**Comment by MattzFiber** [comment]
> @MattzFiber did you ever resolve this issue with Guac version 1.0.0+?

@purplebadger I did not, no, and I need to put more effort into the testing above as requested by @aquesnel 
Have you seen the same condition?

**Comment by purplebadger** [comment]
Yes 100%. I can run Guac v1.0.0 (CentOS) with an RDP connection to Ubuntu (18.04 or 20.04 build with MATE desktop env)  and sound will run seamlessly all day every day. The minute I build a fresh Guac with v1.1.0 (or 1.2.0 or 1.3.0) changing nothing else on the client side, I connect to an RDP desktop session on the Ubuntu machine sound will play for a few seconds then will crash (the remote session continues as normal however without any sound) with:
Dec 31 13:26:10 XXXX xrdp[3050]: (3050)(140335865218880)[DEBUG] Closed socket 18 (AF_UNIX)
Dec 31 13:26:10 XXXX pulseaudio[3230]: [xrdp-sink] module-xrdp-sink.c: data_send: send failed

Like you, I spent many many hours trying to identify the error and finally found the "cause" to be the upgrade to Guacamole. I presumed this was a problem peculiar to my setup as, although not very common, I thought I cant be the only person using Guacamole to connect to an Ubuntu/xRDP desktop environment and play sound - i found this github issue because you have the same symptoms and same log error.
